### PR TITLE
feat(rest): API to get vulnerability tracking status

### PR DESCRIPTION
**Pull request by keerthi-bl**
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? 
#2013 

> * Did you add or update any new dependencies that are required for your change?
No 

End Point : http://localhost:8080/resource/api/vulnerabilities/trackingstatus/8b3aa2f81a764fb482d6391516d153db

Sample Response : 

{
    "_embedded": {
        "sw360:releaseClearingStatusDatas": [
            {
                "release": {
                    "type": "release",
                    "name": "Component2",
                    "version": "1.0",
                    "componentId": "cb38c33187934c85a3c024414ff2dd2a",
                    "releaseDate": "",
                    "externalIds": {},
                    "additionalData": {},
                    "attachments": [
                        {
                            "filename": "projects-2023-03-02.xlsx",
                            "sha1": "349851d0913535925073ca0d34077bd5bd06c002",
                            "attachmentType": "SOURCE",
                            "createdBy": "admin@sw360.org",
                            "createdTeam": "DEPARTMENT",
                            "createdComment": "",
                            "createdOn": "2023-06-15",
                            "checkedComment": "",
                            "checkStatus": "NOTCHECKED"
                        }
                    ],
                    "createdOn": "2023-01-25",
                    "mainlineState": "OPEN",
                    "clearingState": "NEW_CLEARING",
                    "externalToolProcesses": [],
                    "createdBy": "admin@sw360.org",
                    "contributors": [],
                    "subscribers": [],
                    "roles": {},
                    "mainLicenseIds": [],
                    "otherLicenseIds": [],
                    "vendor": {
                        "type": "vendor",
                        "url": "http://localhost:8999",
                        "shortName": "DV",
                        "fullName": "Demo Vendor"
                    },
                    "languages": [],
                    "operatingSystems": [
                        "Linux"
                    ],
                    "softwarePlatforms": [],
                    "sourceCodeDownloadurl": "",
                    "binaryDownloadurl": "",
                    "externalToolProcessesIterator": [],
                    "cpeId": "",
                    "eccInformation": {
                        "eccStatus": "IN_PROGRESS"
                    }
                },
                "componentType": "SERVICE",
                "projectNames": "Demo Project (11)",
                "mainlineStates": "Open",
                "accessible": true,
                "setMainlineStates": true,
                "setProjectNames": true,
                "setComponentType": true,
                "setRelease": true,
                "setAccessible": true
            }
        ]
    },
    "_links": {
        "curies": [
            {
                "href": "http://localhost:8080/resource/docs/{rel}.html",
                "name": "sw360",
                "templated": true
            }
        ]
    }
}

### Checklist
Must:
- [ ] All related issues are referenced in commit messages and in PR

### CODE REVIEW

**Comment by keerthi-bl** [comment]
As per the discussion, the response structure has been changed as follows and committed the changes.

Sample response : 
[
    {
        "releaseid": "e397e509eb2344e28d11898fea168bb0",
        "name": "Component2",
        "projectorigin": "Demo Project (11)",
        "svmtrackingstatus": null,
        "shortstatus": null,
        "type": "SERVICE"
    }
]

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 1]
```diff
@@ -359,4 +365,38 @@ private Set<Release> getReleaseRelationsInfo(Vulnerability vulnerability, User u
 
         return releaseList;
     }
+
+    @GetMapping(value = VULNERABILITIES_URL + "/trackingstatus" + "/{projectid}")
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:369]: change

```
trackingstatus -> trackingStatus
projectid -> projectId
```

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 2]
```diff
@@ -359,4 +365,38 @@ private Set<Release> getReleaseRelationsInfo(Vulnerability vulnerability, User u
...
+    public void getVulnerabilitiesTrackingStatus(@PathVariable("projectid") String projectId,
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:370]: use **ResponseEntity** and avoid **HttpServletResponse**

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 3]
```diff
@@ -359,4 +365,38 @@ private Set<Release> getReleaseRelationsInfo(Vulnerability vulnerability, User u
...
+                                    && !rel.getRelease().getAdditionalData().equals("")
+                                            ? rel.getRelease().getAdditionalData().get("svmShortStatus")
+                                            : "");
+                    jobj.addProperty("type", rel.getComponentType().toString());
+                    resultJson.add(jobj);
+                }
+            });
+            response.getWriter().write(resultJson.toString());
+        } catch (SW360Exception e) {
+            throw new TException(e.why);
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:398]: Handle exception

**Comment by hoangnt2** [comment]
@keerthi-bl, If you can please ignore the unnecessary fields in response

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/Sw360VulnerabilityService.java [hunk 4]
```diff
@@ -306,4 +309,10 @@ public ObjectMapper getObjectMapper() {
     public enum VulnerabilityOperation {
         CREATE, UPDATE, DELETE;
     }
+
+    public List<ReleaseClearingStatusData> getReleaseClearingStatusesWithAccessibility(User user, String projectId)
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/Sw360VulnerabilityService.java:313]: `getReleaseClearingStatusesWithAccessibility` -> `getReleasesClearingStatusWithAccessibility`

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 5]
```diff
@@ -359,4 +365,38 @@ private Set<Release> getReleaseRelationsInfo(Vulnerability vulnerability, User u
...
+            HttpServletResponse response) throws TException, IOException {
+        List<JsonObject> resultJson = new ArrayList<>();
+        try {
+            User user = restControllerHelper.getSw360UserFromAuthentication();
+            List<ReleaseClearingStatusData> releaseClearingStatusList = vulnerabilityService
+                    .getReleaseClearingStatusesWithAccessibility(user, projectId);
+            releaseClearingStatusList.forEach(rel -> {
+                if (rel.getRelease() != null) {
+                    JsonObject jobj = new JsonObject();
+                    jobj.addProperty("releaseid", rel.getRelease().getId());
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:380]: releaseid -> releaseId
projectorigin -> projectOrigin
svmtrackingstatus -> svmTrackingStatus
shortstatus -> shortStatus

File: rest/resource-server/src/docs/asciidoc/vulnerabilities.adoc [hunk 6]
```diff
@@ -125,4 +125,12 @@ A `DELETE` request to delete a release vulnerability relation
...
 
 ===== Example response
-include::{snippets}/should_document_delete_release_vulnerability_relation/http-response.adoc[]
\ No newline at end of file
+include::{snippets}/should_document_delete_release_vulnerability_relation/http-response.adoc[]
+
+[[resources-vulnerabilitie-get-status]]
+==== Get a vulnerability tracking status
+
+A `GET` request will get vulnerability tracking status.
```
Comment (hoangnt2) [inline rest/resource-server/src/docs/asciidoc/vulnerabilities.adoc:133]: Make the description more clear

Example: vulnerability tracking status for a project (or something)

**Comment by keerthi-bl** [comment]
Review comments addressed.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 7]
```diff
@@ -359,4 +365,40 @@ private Set<Release> getReleaseRelationsInfo(Vulnerability vulnerability, User u
...
+                                            ? rel.getRelease().getAdditionalData().get("svmShortStatus")
+                                            : "");
+                    jobj.addProperty("type", rel.getComponentType().toString());
+                    resultJson.add(jobj);
+                }
+            });
+            response.getWriter().write(resultJson.toString());
+        } catch (SW360Exception e) {
+            throw new TException(e.why);
+        }catch(Exception ex) {
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:399]: add space after }

Comment (keerthi-bl) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java]: Resolved

**Comment by keerthi-bl** [comment]
Review comment addressed.

**Comment by hoangnt2** [comment]
@keerthi-bl, the response body doesn't seem to be in JSON format
image

You can try to rewrite the controller like below

```
    @GetMapping(value = VULNERABILITIES_URL + "/trackingStatus" + "/{projectId}")
    public ResponseEntity<List<Map<String, String>>> getVulnerabilitiesTrackingStatus(@PathVariable("projectId") String projectId) throws TException {
        List<Map<String, String>> result = new ArrayList<>();
        try {
            User user = restControllerHelper.getSw360UserFromAuthentication();
            List<ReleaseClearingStatusData> releaseClearingStatusList = vulnerabilityService
                    .getReleasesClearingStatusWithAccessibility(user, projectId);
            releaseClearingStatusList.forEach(rel -> {
                if (rel.getRelease() != null) {
                    Map<String, String> vulnerabilityTrackingStatus = new TreeMap<>();
                    vulnerabilityTrackingStatus.put("releaseId", rel.getRelease().getId());
                    vulnerabilityTrackingStatus.put("name", rel.getRelease().getName());
                    vulnerabilityTrackingStatus.put("projectOrigin",  rel.getProjectNames());
                    vulnerabilityTrackingStatus.put("svmTrackingStatus",
                            null != rel.getRelease().getExternalIds() && !rel.getRelease().getExternalIds().isEmpty()
                                    ? rel.getRelease().getExternalIds().get("svmComponentId")
                                    : "");
                    vulnerabilityTrackingStatus.put("shortStatus",  null != rel.getRelease().getAdditionalData()
                            && !rel.getRelease().getAdditionalData().isEmpty()
                            ? rel.getRelease().getAdditionalData().get("svmShortStatus")
                            : "");
                    vulnerabilityTrackingStatus.put("type",   rel.getComponentType().toString());

                    result.add(vulnerabilityTrackingStatus);
                }
            });
            return ResponseEntity.ok(result);
        } catch (SW360Exception e) {
            throw new TException(e.why);
        } catch(Exception ex) {
            throw new TException(ex.getMessage());
        }
    }
```

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 8]
```diff
@@ -359,4 +365,40 @@ private Set<Release> getReleaseRelationsInfo(Vulnerability vulnerability, User u
...
+            List<ReleaseClearingStatusData> releaseClearingStatusList = vulnerabilityService
+                    .getReleasesClearingStatusWithAccessibility(user, projectId);
+            releaseClearingStatusList.forEach(rel -> {
+                if (rel.getRelease() != null) {
+                    JsonObject jobj = new JsonObject();
+                    jobj.addProperty("releaseId", rel.getRelease().getId());
+                    jobj.addProperty("name", rel.getRelease().getName());
+                    jobj.addProperty("projectOrigin", rel.getProjectNames());
+                    jobj.addProperty("svmTrackingStatus",
+                            null != rel.getRelease().getExternalIds() && !rel.getRelease().getExternalIds().equals("")
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:384]: InternalIds  and AddditionalData is a Map so it is always not equal to "". Use isEmpty()

**Review by hoangnt2 — CHANGES_REQUESTED** [review]

**Comment by keerthi-bl** [comment]
Review comments addressed.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 9]
```diff
@@ -15,14 +15,17 @@
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
+import java.io.IOException;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:18]: remove unused import

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 10]
```diff
@@ -15,14 +15,17 @@
...
 import java.net.URI;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.HashSet;
 import java.util.Set;
+import java.util.TreeMap;
 import java.util.stream.Collectors;
 
+import javax.servlet.http.HttpServletResponse;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:28]: remove unused import

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java [hunk 11]
```diff
@@ -56,17 +60,20 @@
...
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
 
 import java.util.ArrayList;
 import java.util.List;
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.gson.JsonObject;
```
Comment (hoangnt2) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/vulnerability/VulnerabilityController.java:76]: remove unused import

**Review by hoangnt2 — CHANGES_REQUESTED** [review]

**Comment by keerthi-bl** [comment]
Review comments addressed.

**Review by hoangnt2 — APPROVED** [review]
Changes look good

**Pull request merged**